### DISCUSSION TITLE
The internal of converting onnx file to mlir file

### DISCUSSION THREAD

**Discussion by FlintWangacc** [onnx/onnx-mlir#2894] [General]
I'm using onnx-mlir to compile neural network to mlir. And then use iree-compile to lower mlir to vmfb file which runs on RISCV devices. Now I'm quite interested in the process that change onnx to mlir format. Is there some tutorial I can follow or some simple test case which can demonstrate this process? Thank you.

**Comment by AlexandreEichenberger** [comment]
We don't have a manual per se. but the flow is intake onnx to a basic onnx, perform onnx shape inference, perform constant propagation/other operation substitution (e.g. convert old onset into newer ones), and then we start lowering to krnl->affine->llvm.

You can see the process using a simple onnx file, e.g. add.mlir
```
module {
  func.func @main_graph(%arg0: tensor<16x8x128xf32> {onnx.name = "x"}) -> (tensor<16x8x128xf32> {onnx.name = "y"}) {
    %0 = "onnx.Add"(%arg0, %arg0) : (tensor<16x8x128xf32>, tensor<16x8x128xf32>) -> tensor<16x8x128xf32>
    return %0 : tensor<16x8x128xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph} : () -> ()
}
```

and then compiling with flags that show you each step along the way:
```
onnx-mlir -O3 -march=arm64 -mlir-print-ir-after-all add.mlir
```

Default is to generate binary (`-EmitLib`) but do a `onnx-mlir -h` and you will see there are other targets as well.
